File: Graph Theory Peaks/BFS/Q5. Count the Number of Complete Components/Q5_CounttheNumberofCompleteComponents.py
class Solution(object):
    def countCompleteComponents(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: int
        """
        
        from collections import deque
        
        # Create adjacency list to represent the graph
        adj = [[] for _ in range(n)]
        
        # Build the undirected graph
        for u, v in edges:
            adj[u].append(v)   # add v as neighbor of u
            adj[v].append(u)   # add u as neighbor of v
        
        # Track visited nodes to avoid processing them again
        vis = [False] * n
        
        # Store number of complete components
        ans = 0
        
        # BFS function to find all nodes in the current connected component
        def bfs(node):
            q = deque([node])     # queue for BFS traversal
            vis[node] = True      # mark starting node as visited
            comp = []             # list to store nodes of this component
            
            while q:
                cur = q.popleft()  # get next node from queue
                comp.append(cur)   # add node to component list
                
                # Explore all neighbors of current node
                for neighbor in adj[cur]:
                    if not vis[neighbor]:       # if neighbor not visited
                        vis[neighbor] = True    # mark it visited
                        q.append(neighbor)      # add it to queue
            
            return comp   # return all nodes belonging to this component
        
        # Traverse all nodes to find different connected components
        for i in range(n):
            if not vis[i]:             # if node not yet visited
                
                comp = bfs(i)          # get all nodes in this component
                
                # Check if the component forms a complete graph
                # In a complete graph each node connects to (size - 1) nodes
                if all(len(adj[node]) == len(comp) - 1 for node in comp):
                    ans += 1           # count this as a complete component
        
        return ans

File: Graph Theory Peaks/BFS/Q5. Count the Number of Complete Components/Q5_CounttheNumberofCompleteComponents.java
import java.util.*;

class Solution {
    public int countCompleteComponents(int n, int[][] edges) {

        // Create adjacency list for graph
        List<List<Integer>> graph = new ArrayList<>();

        // Initialize empty list for each node
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }
        
        // Build the graph (undirected)
        for (int[] edge : edges) {
            graph.get(edge[0]).add(edge[1]); // edge u -> v
            graph.get(edge[1]).add(edge[0]); // edge v -> u
        }
        
        // Track visited nodes
        boolean[] visited = new boolean[n];

        // Count of complete connected components
        int completeComponents = 0;
        
        // Traverse all nodes
        for (int i = 0; i < n; i++) {

            // If node not visited, explore its component
            if (!visited[i]) {

                // Store all nodes in this connected component
                Set<Integer> component = new HashSet<>();

                // Perform DFS to collect nodes of this component
                dfs(i, graph, visited, component);

                // Check if this component forms a complete graph
                if (isComplete(component, graph)) {
                    completeComponents++;
                }
            }
        }
        
        // Return number of complete components
        return completeComponents;
    }
    
    private void dfs(int node, List<List<Integer>> graph, boolean[] visited, Set<Integer> component) {

        // Mark current node as visited
        visited[node] = true;

        // Add node to current component
        component.add(node);

        // Traverse all neighbors
        for (int neighbor : graph.get(node)) {

            // If neighbor not visited, continue DFS
            if (!visited[neighbor]) {
                dfs(neighbor, graph, visited, component);
            }
        }
    }
    
    private boolean isComplete(Set<Integer> component, List<List<Integer>> graph) {

        // Size of current connected component
        int size = component.size();

        // Check every node in the component
        for (int node : component) {

            // In a complete graph, each node must connect to (size - 1) nodes
            if (graph.get(node).size() != size - 1) {
                return false;
            }
        }

        // If all nodes satisfy condition, component is complete
        return true;
    }
}